### ISSUE TITLE
Use same meaning for [filter] and [filter_self] between abilities and specials

### ISSUE THREAD

**Issue by ProditorMagnus** [wesnoth/wesnoth#9240]
### Describe the desired feature

Current ability
- [filter]: [StandardUnitFilter](https://wiki.wesnoth.org/StandardUnitFilter) If the unit owning the ability does not match this filter, the ability will be inactive.
- [filter_self]: if the owner of the ability does not match this filter, it will not receive the effects of the ability. [filter_self] takes a [StandardUnitFilter](https://wiki.wesnoth.org/StandardUnitFilter) as argument.

Current special
- [filter_self]: the special will only be active if the owner matches this [StandardUnitFilter](https://wiki.wesnoth.org/StandardUnitFilter) (SUF).

I suspect something of documentation might not be accurate anymore, not tested yet.

I think special too should use meanings
- [filter]: If the unit owning the special does not match this filter, the special will be inactive.
- [filter_self]: if the owner of the special does not match this filter, it will not receive the effects of the special

**Comment by newfrenchy83** [comment]
i don't agree because in special exist other filter what filter_self , [filter_oppoent],[filter_attacker] and [filter_defender] and if these filter don't match special inactive then why it should be diffrent with  [filter_self]?

**Comment by ProditorMagnus** [comment]
Example
```
[damage]
	add=1
	apply_to=self,opponent
	[filter]
		side=1
	[/filter]
	[filter_self]
		level=1
	[/filter_self]
	[filter_opponent]
		level=1
	[/filter_opponent]
[/damage]
```
For unit with special with
- side 1 level 1 - should show as active on sidebar
- side 1 level 2 - should show as inactive on sidebar
- side 2 - should show as inactive on sidebar

In attack dialog between
- side 1 level 1 with ability + side 2 level 1 - should show as active on both sides
- side 1 level 2 with ability + side 2 level 1 - should show as active for opponent
- side 1 level 1 with ability + side 2 level 2 - should show as active for self
- side 1 level 2 with ability + side 2 level 2 - should show as inactive on both sides
- side 2 with ability + side 1 - should show as inactive on both sides (neither level matters since entire special is inactive)


This gap is about partially active specials.

**Comment by ProditorMagnus** [comment]
I verified that [filter] does not work in specials, and [filter_self] does not mark special visually inactive. So I think just [filter] should be added to specials.

**Comment by stevecotton** [comment]
@ProditorMagnus in IRC/Discord said:
> I think teaching should be further split - that unit A has a "teaching" ability that confers a chance_to_hit ability to unit B, which gives unit B chance_to_hit special to its all attacks

Thanks for that set of names, I'm using them here:

If `[filter]` was added to specials in the current syntax then in weapon-specials-used-as-abilities it would have to match both A and B, because both the "teaching ability" part and the "special" part would interpret `[filter]`.

**Comment by gfgtdf** [comment]
could someone clarify what the current haviour is, does for example
```[abilities]
  [damage]
    [filter]
      ..
    [/filter]
    ..
  [/damage]
[/abilities]
```

work as expected, so that tge requwat in this issue is only about specials used in `[attack][specials]`?

**Comment by newfrenchy83** [comment]
> could someone clarify what the current haviour is, does for example
> 
> ```
>   [damage]
>     [filter]
>       ..
>     [/filter]
>     ..
>   [/damage]
> [/abilities]
> ```
> 
> work as expected, so that tge requwat in this issue is only about specials used in `[attack][specials]`?

the filter regulates the activity of [damage] when implemented in [abilities] in the same way as for a general ability, but does not work in a [damage] implemented in [special] where we use [filter_self/opponent/attacker/defender], the wikia seems correct on this point and using [filter_self] in a [damage] would cause bugs if it had both meanings at the same time, hence the need to replace [filter_self] with filter_student for the meaning of special implemented in [abilities] (I never dared to try filter_self directly because the problem seemed quite clear to me without testing it).

The problem comes from using filter_self in a capability to use as a special, if [filter_self] is used in this context, it will be the capability meaning of the filter that is exercised and that is why the suggestion here is a bad idea

**Comment by newfrenchy83** [comment]
> [@ProditorMagnus](https://github.com/ProditorMagnus) in IRC/Discord said:
> 
> > I think teaching should be further split - that unit A has a "teaching" ability that confers a chance_to_hit ability to unit B, which gives unit B chance_to_hit special to its all attacks
> 
> Thanks for that set of names, I'm using them here:
> 
> If `[filter]` was added to specials in the current syntax then in weapon-specials-used-as-abilities it would have to match both A and B, because both the "teaching ability" part and the "special" part would interpret `[filter]`.

[filter] works for the 'teaching' part, and so is A doesn't match, the ability is inactive and nothing is taught to B, but why should BE also match for it to work?
